Add utility to test if autograd was registered correctly

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106558
* __->__ #106561

See docstring for more details. This API is not meant to be directly
user-facing, it is meant to be used as a subtest of D47965247, which is
coming soon.

Test Plan:
- Some new tests.

zou3519 committed:
Add utility to test if autograd was registered correctly

See docstring for more details. This API is not meant to be directly
user-facing, it is meant to be used as a subtest of D47965247, which is
coming soon.

Test Plan:
- Some new tests.

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/optests/autograd_registration.py:
```diff
@@ -0,0 +1,135 @@
+      and you wish the operator to decompose and get autograd support
+      that way, then please register the implementation to
+      DispatchKey::CompositeImplicitAutograd
+    - If you're adding an autograd formula for the operator, the correct
```
Should we recommend the custom op API instead/too?

zou3519 replied:
In the future, when the custom op guide is done, I am going to link it from everywhere. I haven't written about the autograd API yet so it wasn't mentioned here

zou3519 committed:
Update on "Add utility to test if autograd was registered correctly"


See docstring for more details. This API is not meant to be directly
user-facing, it is meant to be used as a subtest of D47965247, which is
coming soon.

Test Plan:
- Some new tests.

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/optests/autograd_registration.py:
```diff
@@ -0,0 +1,135 @@
+    # responsibility to a different test (schema_check).
+
+    # If no inputs require grad, we are unable to perform this test, so we
+    # just succeed.
```
Maybe we should error instead?

zou3519 replied:
It works a bit better for the crossref testing if this doesn't actually error, but if we feel strongly about the error I can make it error with a special error and then surpress it during crossref testing.

soulitzer replied:
My vote would be for erroring and then suppressing where necessary

soulitzer replied:
Oh I just checked, gradcheck actually errors if no inputs require grad! so probably good to do the same here.

zou3519 replied:
Sounds good!

soulitzer commented on torch/testing/_internal/optests/autograd_registration.py:
```diff
@@ -0,0 +1,135 @@
+    #   grad. This would only happen if the user calls regular PyTorch
+    #   operations in their backend key (this op should instead be
+    #   CompositeImplicitAutograd or not an op) or if the user invokes
+    #   an autograd.Function in the backend key.
```
What if the op is partially a composite op and partially does non-pytorch operations? Are we sure we also want to error in that case? I guess in theory this means they could extract out a smaller custom operation. I'm not sure how often this happens in practice though.

If we wanted to test that the operator is "fully" composite that would probably mean having to actually validate the gradients are wrong, but maybe that is too complex to be worth?

zou3519 replied:
> What if the op is partially a composite op and partially does non-pytorch operations? Are we sure we also want to error in that case?

Yup. Ideally the user makes the custom op as small as possible.

> If we wanted to test that the operator is "fully" composite that would probably mean having to actually validate the gradients are wrong, but maybe that is too complex to be worth?

Yeah, this is meant to be a simple test.

soulitzer replied:
Ah makes sense. I don't know how common it is for folks to mix normal torch ops in with their custom ops, but if it is, it might be worth mentioning this case in the error/docs.

zou3519 committed:
Update on "Add utility to test if autograd was registered correctly"


See docstring for more details. This API is not meant to be directly
user-facing, it is meant to be used as a subtest of D47965247, which is
coming soon.

Test Plan:
- Some new tests.

[ghstack-poisoned]

zou3519 committed:
Update on "Add utility to test if autograd was registered correctly"


See docstring for more details. This API is not meant to be directly
user-facing, it is meant to be used as a subtest of D47965247, which is
coming soon.

Test Plan:
- Some new tests.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
